src/api.js

```
@@ -0,0 +1,110 @@
+        return {
+          lastModified: newLastModified,
+          changes: json.items
+        };
+      });
+  }
+
+  batch(collName, records, headers={}, options={safe: true}) {
+    const results = {
+      errors:    [],
+      published: [],
+      conflicts: [],
+    };
+    var reject = false;
+    if (!records.length)
+      return Promise.resolve(results);
+    return fetch(this.endpoints().batch(), {
+      method: "POST",
+      headers: DEFAULT_REQUEST_HEADERS,
+      body: JSON.stringify({
+        defaults: {
+          headers: headers,
+        },
+        requests: records.map(record => {
+          const isDeletion = record._status === "deleted";
+          const path = this.endpoints({full: false}).record(collName, record.id);
+          const method = isDeletion ? "DELETE" : "PUT";
+          const body = isDeletion ? undefined : cleanRecord(record);
+          const headers = options.safe && record.last_modified ?
+                          {"If-Unmodified-Since": String(record.last_modified)} : {};
+          return {method, headers, path, body};
+        })
+      })
+    }).then(res => {
+      if (res.status === 400)
+        throw new Error("Invalid BATCH request"); // TODO: precise error reporting
+      if (res.status !== 200)
+        throw new Error("BATCH request failed, HTTP " + res.status); // TODO: precise error reporting
+      return res.json();
+    }).then(res => {
+      res.responses.forEach(response => {
+        if (response.status && response.status >= 200 && response.status < 400) {
+          results.published.push(response.body);
+        } else if (response.status === 412) {
+          results.conflicts.push(response.body);
+          reject = true;
+        } else {
+          results.errors.push({
+            path: response.path, // this is the only way to have the id…
+            error: response.body
+          });
+          reject = true;
+        }
+      });
+      if (reject)
+        throw results;
```

throwing in a then()able needs a catch() method as well, or the error will get silently ignored iirc